src/librustc_privacy/lib.rs

```diff
@@ -257,83 +305,51 @@ impl<'a, 'tcx, 'v> Visitor<'v> for EmbargoVisitor<'a, 'tcx> {
-            hir::ItemImpl(_, _, _, _, ref ty, ref impl_items) => {
-                let public_ty = match ty.node {
-                    hir::TyPath(..) => {
-                        match self.tcx.def_map.borrow().get(&ty.id).unwrap().full_def() {
-                            def::DefPrimTy(..) => true,
-                            def::DefSelfTy(..) => true,
-                            def => {
-                                let did = def.def_id();
-                                if let Some(node_id) = self.tcx.map.as_local_node_id(did) {
-                                    self.exported_items.contains(&node_id)
-                                } else {
-                                    true
-                                }
-                            }
-                        }
-                    }
-                    _ => true,
-                };
-                let tr = self.tcx.impl_trait_ref(self.tcx.map.local_def_id(item.id));
-                let public_trait = tr.clone().map_or(false, |tr| {
-                    if let Some(node_id) = self.tcx.map.as_local_node_id(tr.def_id) {
-                        self.exported_items.contains(&node_id)
-                    } else {
-                        true
-                    }
-                });
+            hir::ItemImpl(_, _, _, Some(ref trait_ref), ref ty, ref impl_items) => {
+                let (public_ty, _exported_ty) = self.is_public_exported_ty(&ty);
+                let (public_trait, exported_trait) = self.is_public_exported_trait(trait_ref);
 
-                if public_ty || public_trait {
-                    for impl_item in impl_items {
-                        match impl_item.node {
-                            hir::ConstImplItem(..) => {
-                                if (public_ty && impl_item.vis == hir::Public)
-                                    || tr.is_some() {
-                                    self.exported_items.insert(impl_item.id);
-                                }
-                            }
-                            hir::MethodImplItem(ref sig, _) => {
-                                let meth_public = match sig.explicit_self.node {
-                                    hir::SelfStatic => public_ty,
-                                    _ => true,
-                                } && impl_item.vis == hir::Public;
-                                if meth_public || tr.is_some() {
-                                    self.exported_items.insert(impl_item.id);
-                                }
-                            }
-                            hir::TypeImplItem(_) => {}
-                        }
-                    }
+                if public_ty && public_trait { self.public_items.insert(item.id); }
+                if exported_trait { self.exported_items.insert(item.id); }
+
+                for impl_item in impl_items {
+                    if public_ty && public_trait { self.public_items.insert(impl_item.id); }
```

Comment:
@arielb1 
That line was indeed at fault, but it was mitigated by rustc_privacy. (The "undefined reference" in my example wasn't an actual error, it's from my temporary branch where "private trait for private type" impls are not exported)

By the way, does rustc have a function like `get_impl_method_id_if_it_is_available_without_doing_monomorphization(MethodCall) -> Option<ImplOrTraitItem>`?
Example:
```
trait Tr {
    fn method(&self) {} // ID_1 (default impl)
}

impl Tr for u8 {}
impl Tr for u16 {
    fn method(&self) {} // ID_2
}

fn generic<T: Tr>(arg: T) {
    0u8.method(); // concrete context, get_impl_method_id_... returns Some(ID_1)
    0u16.method(); // concrete context, get_impl_method_id_... returns Some(ID_2)
    arg.method(); // generic context, get_impl_method_id_... returns None
}
```